### DISCUSSION TITLE
TCA showcase

### DISCUSSION THREAD

**Discussion by shahzadmajeed** [pointfreeco/swift-composable-architecture#1145] [Q&A]
Hello,

One of the data point we are missing during TCA evaluation is existing apps that are built with TCA so we can try them out and see their complexity and user experience. This will make us more confident about using TCA for our future and current (long term large scale) apps.

TCA Authors: Do we have such a list?
Developers:  If you have built apps with TCA and can include the name and link to App Store in the discussion below that would be greatly appreciated

Thank you

**Comment by saroar** [comment]
here is my own two app you can 
https://github.com/AddaMeSPB/AddaMeIOS -> https://apps.apple.com/app/id1538487173
https://apps.apple.com/us/app/word300/id1619504857

**Reply by shahzadmajeed** [reply]
Thank you for sharing these

**Comment by josephktcheung** [comment]
I've scanned through Github and here's a list of open sourced apps that uses TCA:

[Puffery](https://github.com/vknabel/puffery) App store: https://apps.apple.com/de/app/puffery/id1508776889
[reminders-app](https://github.com/lukeleleh/reminders-app)
[Marconio](https://github.com/brianmichel/Marconio)
[Prose macOS](https://github.com/prose-im/prose-app-macos)
[DroidKaigi 2021 iOS Application](https://github.com/DroidKaigi/conference-app-2021/tree/main/ios)
[weg-li](https://github.com/weg-li/weg-li-ios)
[Seed](https://github.com/lcr3/Seed)
[runitz](https://github.com/jjorgemoura/runitz)
[Critical Maps iOS App](https://github.com/criticalmaps/criticalmaps-ios) iTunes: https://itunes.apple.com/app/critical-maps/id918669647
[GuitarTuner](https://github.com/kodydeda4/SwiftUI-GuitarTuner)
[Tube service](https://github.com/ridgeview-apps/tube-service) App store: https://apps.apple.com/app/id1540626286#?platform=iphone

**Reply by shahzadmajeed** [reply]
Thank you for this list ❤️

**Comment by pacu** [comment]
Our App is still in the making but it's SwiftUI + TCA.

https://github.com/zcash/secant-ios-wallet

I've developed one of the first SwiftUI apps in the crypto space back in 2019 when SwiftUI was a mere beta. I can tell you that Vanilla Swift UI doesn't work for not even for "the Fruta App" applications. You need more tools. TCA is not a silver bullet but a very flexible tool that provides a framework that a team can use to build an application that uses unidirectional data driven flow to drive SwiftUI's DSL. 

My main concern when evaluating TCA for building secant, was the learning curve. If your project is a fully open, everyone can contribute freely kind of project, TCA could be either: 1) a very attractive honey pot to get developers the TCA crave they were seeking, or 2) a contributor barrier. Only time will tell, currently is a coin-toss. If your project is closed-source and 9-to-5 coding. Your team will be very happy of learning new stuff. Every new hire of the future will benefit of having a clear onboarding route already created for you by the pointfree dudes.

What I can say about my experience so far:

**The CONs:**

- TCA has a learning curve proper of using a new paradigm with familiar tools.  **Takeaway**: you will have this learning curve whatever your choice is if you are not familiar with it.
- You will have to create wrappers for dependencies you once used innocently "out-of-the-box". **Takeaway:** this was probably an illusion. You are paying the cost up front and before you were deferring it until it blew up on your face. 
- Not the Drag and Drop solution you dreamed of.

**The PROs:**
- it's is very well documented. But, you have to make the time for reading it and sink it in. Unless you already know the functional paradigm, there are things than won't be easy so figure out just by cmd+click and reading the code. 
- it provides a very opinionated way of handling state and data flow on your app. This reduces SwiftUI's bugs significantly. SwiftUI makes it easy to couple screens to one another. Combine observers can drag down your app's performance due to unexpected re-renders and hierarchy recreations that will make your app flicker, reset at best if not crash. 
- it makes you have testable code by default. Having to wrap your dependencies in witnesses or wrappers lets you mock them you will have `.live` and `.demo` implementations of every dependency you once used to struggle with. Your app is far closer to 100% test coverage than ever before.
- it makes you a better developer. There's no way you go though the docs without learning something new. 
- using `Store` and `Reducer` make your business logic fully unit-testable. 

**other comments:**
You will compensate the learning curve with time saved by not having to deal with bugs caused by mishandled bindings, random re-renders, etc.

**How I approached it:**

The way I found to approaching TCA is actually subscribing to pointfree.co content and study it. Try to rely on people that has experience on it to help your team gain traction and experience. If not, create study groups and dedicate time of your day into absorbing the content generated by the authors. You won't be able to absorbe more than 45 minutes of content a day. Since there are many concepts that are covered throughout the episodes and in the end the only way to learn them fully is to apply them. If you have an urge, try to find contractors that are very familiar with it and let them help you ease the curve.

**Reply by jshier** [reply]
> * TCA has a learning curve proper of using a new paradigm with familiar tools. Takeaway: you will have this learning curve whatever your choice is if you are not familiar with it.

I've found this learning curve always exists, even if you write your own app architecture from scratch. It's just that, by writing the architecture from scratch, you're learning the architecture while writing it, which will likely take more time overall than simply learning an architecture. If you take external contributions, contributors always have to ramp up on the architecture. At least with TCA there's extensive documentation, videos, and others using it, rather than none of those things for a bespoke solution.

**Reply by pacu** [reply]
Plus, there's a big chance that writing your own architecture goes wrong and you end up (luckily) re-inventing the wheel. If you have good fortune our wheel will turn without squiggling much.

**Reply by stephencelis** [reply]
@pacu Thanks for the feedback! We hope that some of the updates we have planned will help with some of the learning curve. We should probably get a proper "beta" channel for TCA going so folks can play with some of these planned features a bit earlier so that they can provide feedback. We do a lot of prolonged testing and refinement before releasing new features to TCA, but the more folks that have hands on it the better. Would you be interested in such a channel?

**Reply by shahzadmajeed** [reply]
Hi,

Thank you for sharing your input. I love this discussion. I cannot share exact details of our assessment right now but here are some points:

**TCA**
- ~30-40% of TCA code was needed to build features (local store, state and environment). We didn't want to use it for everything i.e., single store and pullback etc..
- Explicit thread safety is needed to avoid dirty reads & crashes i.e., all store.send calls should be on same thread (main)
- Learning curve was higher but I 💯 agree that learning curve is involved with every new architecture/pattern 
- Testability is amazing with TCA, specially being able to test your logic with just a mock function (effect) was mind blowing 
- There were lot of other factors considered which I'm skipping

Conclusion:
We will write our own lightweight (~200 LOC) state machine with some Combine extensions and use that within a home-written VIPER stack. This will allow us to extend our architecture to meet our uses cases/needs.

**Comment by gonzalonunez** [comment]
We're big fans of TCA at @withprimer :)

There is definitely a learning curve for new engineers, but the pros far outshine the cons:

- Testability is unlike any other iOS app architecture I've ever seen, the cost of writing and maintaining _good_ tests is extremely low
- It increases the ease with which you can interop between SwiftUI and UIKit (passing `Store`s around is simple) which has been really nice for us
- It's fun to use and write code with! There is a fair amount of boilerplate for new features but these are solvable problems with increasingly mature tooling

You can download the Primer app [here](https://primer.com/download). We also have a second smaller app, [Zines](https://apps.apple.com/us/app/zines-share-your-work/id1616263431), that is also built with TCA (and free to use, so it's probably more useful to check out).

**Reply by shahzadmajeed** [reply]
Interop is an interesting point and I didn't think if it before. I can definitely see how TCA can help with that. Did you face any issue with thread safety?

**Reply by gonzalonunez** [reply]
No! At least none that I'm aware of and we monitor our errors well. For what it's worth, we do very basic multithreading, nothing crazy.

**Comment by ZevEisenberg** [comment]
We're using TCA in the new how-to app for kids at The New York Times. We're loving the increased testability it affords.

**Reply by shahzadmajeed** [reply]
Thank you for your input. Which architecture do you use for main apps at NYTimes?

**Reply by shahzadmajeed** [reply]
Did you face any issue with thread safety?

**Reply by ZevEisenberg** [reply]
The main news app uses flow controllers and UIKit. Haven't had many thread safety issues that I can recall, but our threading needs are pretty simple, and I imagine they'll get simpler with new concurrency things.

**Reply by krzysztofzablocki** [reply]
main apps were MVVM+C, TCA made development really fast

**Comment by simme** [comment]
I'm using TCA for my two apps [Handla](https://handlagrocerylist.app) and [Plantry](https://www.plantry.app). The latter is only partly converted thus far. The former was built from the ground up using TCA.

Converting an old codebase has been pretty straightforward, started at the leaf nodes and new features. Working my way inwards.

Building a new app from scratch with TCA was a total joy! Even when building with UIKit I was able to get up and running quickly because it makes mocking data and functionality _so easy_!

Every time I work on an app that doesn't use TCA I miss it!

**Comment by erikolsson** [comment]
I’ve used TCA for a bunch of different apps. My latest one is called Effekt, built on TCA + UIKit + Metal:
https://apps.apple.com/us/app/effect-live-wallpaper/id1616146386
![CF2429F2-DF95-4524-8B32-444C26A94FE7](https://user-images.githubusercontent.com/352982/174956371-e370e39a-69fc-407f-bb5d-7014a6b13339.png)

**Comment by rcarver** [comment]
The upcoming 2.0 release of my app [Series](https://apps.apple.com/us/app/id1471473571) is TCA + SwiftUI. It's in TestFlight now, happy to add you.

I've been using TCA for just about a year now, I converted a client's app from vanilla SwiftUI and then decided to use it for my own app rewrite. Overall it's saved me way more headaches than its created. When I follow the patterns things tend to go very well.

In that spirit, my 2c on making TCA more accessible is to define and document common patterns for using it. For me, the hardest part is designing for composition and composing state (sharing state, delegating effects to parent). There's so much knowledge  in the Point Free videos, isowords and the case studies but it's hard to get a quick tip when you're up against a new problem.

**Reply by ghost** [reply]
> In that spirit, my 2c on making TCA more accessible is to define and document common patterns for using it. For me, the hardest part is designing for composition and composing state (sharing state, delegating effects to parent). There's so much knowledge in the Point Free videos, isowords and the case studies but it's hard to get a quick tip when you're up against a new problem.

I wish I could upvote this comment a million times.

**Comment by jdowd7** [comment]
We are using TCA @ [Exer Labs](https://www.exer.ai/)- its the basis of our 2.0 [Gait Application](https://www.exer.ai/gait).
We focus on providing solutions in the Physical Therapy and Elder Care space with machine learning and augmented reality all on iOS.

<img width="518" alt="image" src="https://user-images.githubusercontent.com/2521560/175393109-a6637f0a-d050-40b0-a652-9be2d3b13b2e.png">

**Comment by romero-ios** [comment]
My budgeting app, Money Flow, was re-written from MVVM to TCA for iOS 15 - https://www.moneyflow.app

**Comment by junjielu** [comment]
Our app [PDB](https://apps.apple.com/app/id1614566357) is building entirely on TCA + UIKit.
Thank you so much for this extraordinary framework!

**Comment by zboralski** [comment]
We just got approval for our new app that uses TCA + SwiftUI. It is really a personal data management app (starting with the Contacts domain). The user interface is completely content-agnostic and everything from the user interface perspective is a graph of propositions you can interact with based on each proposition's type.

https://apps.apple.com/app/id1638189843

Manage and present your contacts with the [Belua](https://apps.apple.com/app/id1638189843) app.

### Features:

### Organise
- Display generative art for contacts without photo.
- Use tags to categorise your contacts.
- Use the touch action to surface contacts in recent.
- Sort by recent, by recently added, by tag or by country.
- Privacy Built-In from the start. 
- Actions such as favourite, tag and hide work across devices if you are using iCloud.
- Works in Airplane mode 

Search
- Search text in contacts
- Diacritic insensitive [a diacritic is a glyph added to a letter or to a basic glyph]. 
- Works with natural language text to ensure correct behaviour in multiple scripts and languages such as Chinese or Japanese.

The app also features privacy built-in from the start, so you can feel confident that your information is safe.

<img src="https://user-images.githubusercontent.com/104316/183397951-7635318d-cdae-4f33-9804-8321693be09d.png" alt="alt text" width="400px">

<img src="https://user-images.githubusercontent.com/104316/183398024-f968ac4a-6fa7-43b8-a3e3-f93829cd2821.png" alt="alt text" width="400px">

<img src="https://user-images.githubusercontent.com/104316/183398130-beee2cee-c69b-4940-9938-290dbfa5c8de.png" alt="alt text" width="400px">

**Comment by mluisbrown** [comment]
The Trading.com app (https://apps.apple.com/app/id1576478434 - US only) is built entirely using TCA and UIKit. Initially it was developed using the [ReactiveSwift fork of TCA](https://github.com/trading-point/reactiveswift-composable-architecture) as we were targeting iOS 12 as a minimum. However that changed during development and the code was re-factored to use the "real" TCA and Combine.

An interesting point is that the Android version of the app was also developed using an in-house Kotlin port of TCA. 

It's a pretty complex app with a lot of real time information streams and continuous UI updates. If you have access to the US App Store you can try it out without having do any real trading by using a demo account with fake money.

Since then I moved on to work at [Argent](https://www.argent.xyz) and the iOS app is mostly TCA, although there is some older code using an in-house redux style architecture.

Right now I can't imagine myself working on any iOS app not using TCA. The difference it makes is really immense.

**Comment by FredericRuaudel** [comment]
I've rebuild the app I'm in charge at [Ver de Terre Digital](https://verdeterredigital.fr) using TCA and I'm really happy with the result. The development is more flexible, I can easily add new features, tests them, create preview apps to live test them, and simulate screen state to generate my AppStore screenshots using fastlane snapshot ! I highly recommend it ! 💚 

[Soildiag on the AppStore](https://apps.apple.com/app/soildiag/id1550766706)

![iPhone 11 Pro Max-04PlantIdentification](https://user-images.githubusercontent.com/3808140/185580040-cd3ff672-94ac-4dc8-b76f-8d4bc3a46267.png) ![iPhone 11 Pro Max-00Home](https://user-images.githubusercontent.com/3808140/185580047-b2dc769e-3a70-4b2d-b884-087a0b7fa693.png) ![iPhone 11 Pro Max-02DiagnosticReport](https://user-images.githubusercontent.com/3808140/185580050-e141470b-d405-4422-a640-81ff3b8fb6a9.png)

**Comment by jeffersonsetiawan** [comment]
[Tokopedia](https://apps.apple.com/us/app/tokopedia/id1001394201): Indonesia Ecommerce app
It's using [RxSwift](https://github.com/tokopedia/RxComposableArchitecture) version of TCA because we still support iOS < 13 and still using cocoapods.
![Simulator Screen Recording - iPod touch (7th generation) - 2022-08-30 at 16 23 14](https://user-images.githubusercontent.com/14871122/187400734-c1db0d20-5dba-4437-9df6-78ec539a1259.gif)

![Simulator Screen Recording - iPod touch (7th generation) - 2022-08-30 at 16 24 20](https://user-images.githubusercontent.com/14871122/187400963-5fdaabc5-6856-4e72-bf4b-edde41a3a9f3.gif)

Really enjoy using the TCA, highly recommended!

**Reply by saroar** [reply]
wow I am ready for work your company let me know if you need any dev :)

**Comment by andreweades** [comment]
Readable is a small app built with TCA. I've just updated it for iOS16.

![Simulator Screen Shot - iPhone 13 Pro Max - 2022-09-21 at 11 28 12](https://user-images.githubusercontent.com/48011137/191851521-06a1eb52-3cde-4bd2-bade-221e92e74e88.png)


https://apps.apple.com/us/app/readable/id1474243671?platform=iphone

**Comment by kuglee** [comment]
Here's mine:
Swift Formatter: Xcode extension for Apple's swift-format code formatter
https://github.com/kuglee/SwiftFormatter

<img width="752" alt="Névtelen" src="https://user-images.githubusercontent.com/1316211/192215347-39546e95-844c-411d-8284-4ed1fd75b4f6.png">

**Comment by krzysztofzablocki** [comment]
[SourceryPro](https://www.merowing.info/sourcery-pro/) uses TCA too, most powerful Xcode extension that enables live AST based templates

![image](https://user-images.githubusercontent.com/1468993/192239841-752f5892-9190-4f4e-bd02-f7351a486d08.png)

**Comment by byMohamedali** [comment]
Hello, 

Hope you're doing well, here is my [app](https://apps.apple.com/us/developer/mohamed-ali-ben-yaagoub/id1151067698) it's a BIG project, on 3 differents platform (macOS, tvOS and iOS) using the same core code, with TCA and packaging feature like you did on isoword project, it's making my life so easier to maintain the 3 platforms :)

I'm currently starting to migrating to reducer protocol. 

Thank you again,

**Comment by birnimal** [comment]
I have built a music player app [Music Rover](https://apps.apple.com/app/music-rover/id1644883513) with TCA.
SwiftUI and MusicKit.

Like many music player apps, it requires complex navigation back and forth between albums, playlists, and artists, so I used [TCACoordinators](https://github.com/johnpatrickmorgan/TCACoordinators).

I usually use Next.js and Redux for web development, and TCA + SwiftUI is great because it is easy to learn and develop in a similar way to them.

One thing I could not figure out how to do is to send actions from multiple functions to a specific function.
For example, since song playback is performed by various functions, I wanted to send an action from each function to the function that manages playback. However, it seems that in order to send actions across functions, it is necessary to make the reducer a parent-child relationship, and I gave up on this idea because I thought it would complicate the structure. Instead, I now create a Client and make a request to it with a PassthroughSubject.

<img src="https://user-images.githubusercontent.com/3584991/209527731-6d4d6604-dcf2-40a7-8d77-0dfcefdceec2.jpg" width="40%"> <img src="https://user-images.githubusercontent.com/3584991/209528790-671233f7-7dd0-477d-986d-6da1d4462fc3.PNG" width="40%">

**Comment by barksten** [comment]
We have an app that's been built with TCA, SwiftUI and MapKit since the beginning. Unfortunately it's only usable in Sweden.
[Min Karta@appstore](https://apps.apple.com/app/id1495766306)

**Comment by Bartozo** [comment]
Hello 👋

I have created a Life Progress app with help of The Composable Architecture (TCA). It uses the latest ReducerProtocol and many other cool things 🙌 Feel free to check out the repository and let me know what do you think about it 🙂

🔗 Here's the link: 
https://github.com/Bartozo/Life-Progress-iOS

![IMG_2691](https://github.com/pointfreeco/swift-composable-architecture/assets/10104462/788f40cc-1f57-40f3-b7ac-f8e8a28a3022)

**Comment by emrdgrmnci** [comment]
Hi all 👋🏼

I have created a NBA teams and players list app with The Composable Architecture (TCA).

🔗 Here's the link:
https://github.com/emrdgrmnci/MatchScoresTCA

**Comment by Sajjon** [comment]
[Radix DLT's wallet app uses TCA](https://github.com/radixdlt/babylon-wallet-ios)

**Comment by akashsoni01** [comment]
Currency exchange app written in TCA. https://github.com/akashsoni01/currency_exchange_app

**Comment by Echim2016** [comment]
Hi all,

I'm currently launching an engaging icebreaker game app called "Check-in Box" using TCA and currently supports Traditional Chinese only. This app sparks conversations through specific topics and fosters closer connections between you and your friends. 🧋

https://github.com/Echim2016/check-in-box-ios
https://apps.apple.com/us/app/id6474719999

**Comment by rokon-uddin** [comment]
Hi,
At Monstarlab we maintain a TCA-based iOS project starter template. Feel free to have a look and share your feedback:
https://github.com/monstar-lab-oss/swiftui-template-tca

**Comment by ba01ei** [comment]
I migrated a simple app to entirely use TCA, and with strict-concurrency fully enabled too:
https://apps.apple.com/us/app/just-keep-favorite-photos/id6461162504

it's a super simple app (it cleans up your photo library), but still, migrating to TCA is fun :)

**Comment by davidcairns** [comment]
We use TCA for the biggest & most complex parts of [Butterfly iQ](https://apps.apple.com/in/app/butterfly-iq-ultrasound/id1183035589), which is a FDA-approved medical ultrasound that plugs into a iPhone/iPad. We rewrote our whole exam screen (the central & most complex part of our app) in TCA way back in 2020, and since then most new code is written using TCA.

So: it’s not just for word games and to-do apps! 😜 

I would be remiss in not mentioning that we won an Apple Design Award in 2019 😏

**Comment by adriansergheev** [comment]
[geoguessr - for old photos](https://github.com/adriansergheev/photoguessr) in tca 
it was written last year and it is open source since a few months now 🙂

**Comment by uniqby** [comment]
My very first app using TCA – [Bible Reader](https://apple.co/3XuEZ3m)

<a href="https://apple.co/3XuEZ3m"><img src="https://github.com/user-attachments/assets/dac6a5f1-e7ab-49c3-9561-772b71f783c9" width=512 height=512/></a>

**Comment by Raidansz** [comment]
An Audio Player in TCA
https://github.com/Raidansz/AudioPlayer-SwiftUI-TCA

**Reply by gcohen-dev** [reply]
It looks nice, I would export the CurrentValueSubject to a client and use AsyncStream, which make more sense IMO, then you don't need to use ObservationStateIgnored.

**Comment by wshamp** [comment]
https://apps.apple.com/us/app/sprigling/id6569230006

-100% TCA uses firebase auth, Firestore,  Apple Subscriptions, Swift 6 

Also started a youtube channel going over the features and how I did them in TCA. - https://www.youtube.com/@ThicketStudios

**Comment by hpetru** [comment]
https://github.com/github/CopilotForXcode is using TCA, it's a very well written extension btw!

**Comment by hpetru** [comment]
Here is one more: https://github.com/Saik0s/Whisperboard 😊

**Comment by dpitkevics** [comment]
Caliverse  - calisthenics and bodyweight fitness app rewritten from Flutter to Swift + SwiftUI using TCA.
https://apps.apple.com/us/app/caliverse-bodyweight-fitness/id1472004214

**Comment by hanneskaeufler** [comment]
I built an app that is arguably 20 years late to the market: Meet [`GalleryMagic`](https://usegallerymagic.com) for quickly generating an HTML image gallery just by dropping a bunch of images into the app. It has a few of the low-hanging fruit features you might expect, and can also publish galleries straight to the web. So I guess now I am also in the hosting business lol. We'll see ...

Of course built using TCA (second time of using in in an app that I actually shipped, again thoroughly enjoyed it). I have some light usage of `swift-sharing` in there as well, though not much yet, and `swift-sharing` to me also hasn't quite "clicked" yet as the general TCA approach has.

As is typical for me being a "I enjoy the developing part, not so much the marketing part:, market research and investment into publicizing it are at a low bar 😆  I also tried my best to ship it early, so I'm not yet proud of my creation. Bugs and warts can be found!
Anyway, here's the App Store link: https://apps.apple.com/us/app/gallerymagic/id6744827234

**Comment by Echim2016** [comment]
Hi all, 

Drift is a CD-style music player I built in SwiftUI, using TCA and Tuist. 
It’s inspired by the early-2000s CD player vibe, especially for long drives.

It’s my second app built with TCA, and using it again made me appreciate even more how well it supports real-world complexity.

Would love your thoughts 🙌

https://apple.co/4rTcZ7N 

<img width="201" height="437" alt="Simulator Screenshot - iPhone 17 Pro - 2025-12-07 at 23 48 22" src="https://github.com/user-attachments/assets/f645c5e4-15a2-4c8c-aeb2-93163e1b590a" />
